File: src/main/GO/leetcode/lists/offer/Offer16.go
package offer

/**
 * @author wujingxinit@outlook.com
 * @date 2022/12/28 23:32
 */

func myPow(x float64, n int) float64 {
	if x == 0 {
		return x
	}
	//fix n == 0
	if n == 0 {
		return 1
	}
	//fix n < 0
	if n < 0 {
		n = -n
		x = 1 / x
	}
	if n == 1 {
		return x
	}
	if n%2 == 0 {
		return myPow(x*x, n/2)
	}
	return x * myPow(x*x, n/2)
}

File: java/src/leetcode/lists/offer/Offer16.java
package leetcode.lists.offer;

/**
 * @author jingxinwu
 * @date 2021-11-21 2:17 下午
 */
public class Offer16 {

    public double myPow(double x, int n) {
        boolean flag = false;
        if (n < 0) {
            flag = true;
            x = 1 / x;
            n = -(n + 1);
        }

        double res = 1;
        double temp = x;

        while (n != 0) {
            if (n % 2 == 1) {
                res *= temp;
            }
            temp *= temp;
            n = n / 2;
        }

        if (flag) {
            res *= x;
        }

        return res;
    }
}
